Add initial implementation of FileSystemWritableFileStream

szewai opened the pull request:
#### 92cb5665c378bef8d73a61a782ad57291520f77d
<pre>
Add initial implementation of FileSystemWritableFileStream
<a href="https://bugs.webkit.org/show_bug.cgi?id=283841">https://bugs.webkit.org/show_bug.cgi?id=283841</a>
<a href="https://rdar.apple.com/problem/140709733">rdar://problem/140709733</a>

Reviewed by Youenn Fablet.

Add initial support for FileSystemWritableFileStream: client can now create it via createWritable() method and use it to
modify an OPFS file (via WritableStreamDefaultWriter methods) in Window context.
Spec: <a href="https://fs.spec.whatwg.org/#api-filesystemwritablefilestream.">https://fs.spec.whatwg.org/#api-filesystemwritablefilestream.</a>

New tests: storage/filesystemaccess/writable-file-stream-write.html
           storage/filesystemaccess/writable-file-stream-seek.html
           storage/filesystemaccess/writable-file-stream-truncate.html

Rebaseline existing tests since they can proceed futher after createWritable() is available. Some tests still fail
because the methods of FileSystemWritableFileStream are not implemented yet.

* LayoutTests/imported/w3c/web-platform-tests/file-system-access/sandboxed_FileSystemDirectoryHandle-move.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/file-system-access/sandboxed_FileSystemDirectoryHandle-move.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-IndexedDB.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-IndexedDB.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-BroadcastChannel.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-Error.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-MessagePort-frames.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-MessagePort-windows.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-MessagePort-workers.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-frames.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-windows.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-workers.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-remove.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-remove.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-getFileHandle.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-getFileHandle.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-iteration.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-iteration.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-removeEntry.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-removeEntry.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-getFile.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-getFile.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-move.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-move.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-sync-access-handle-writable-lock.https.tentative.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream-piped.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream-piped.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream-write.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream-write.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/idlharness.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/idlharness.https.any.worker-expected.txt:
* LayoutTests/storage/filesystemaccess/resources/shared.js:
(finishTest):
(async asyncReadFileAsText):
* LayoutTests/storage/filesystemaccess/resources/writable-file-stream-seek.js: Added.
(async test):
* LayoutTests/storage/filesystemaccess/resources/writable-file-stream-truncate.js: Added.
(async checkFileContent):
(async test):
* LayoutTests/storage/filesystemaccess/resources/writable-file-stream-write.js: Added.
(input.new.Number):
(async runTest):
(async test):
* LayoutTests/storage/filesystemaccess/writable-file-stream-seek-expected.txt: Added.
* LayoutTests/storage/filesystemaccess/writable-file-stream-seek.html: Added.
* LayoutTests/storage/filesystemaccess/writable-file-stream-truncate-expected.txt: Added.
* LayoutTests/storage/filesystemaccess/writable-file-stream-truncate.html: Added.
* LayoutTests/storage/filesystemaccess/writable-file-stream-write-expected.txt: Added.
* LayoutTests/storage/filesystemaccess/writable-file-stream-write.html: Added.
* Source/JavaScriptCore/runtime/GenericTypedArrayView.h:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::copyFile):
* Source/WTF/wtf/FileSystem.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.cpp:
(WebCore::FileSystemFileHandle::createWritable):
(WebCore::FileSystemFileHandle::closeWritable):
(WebCore::FileSystemFileHandle::executeCommandForWritable):
* Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.h:
* Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.idl:
* Source/WebCore/Modules/filesystemaccess/FileSystemStorageConnection.h:
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.cpp: Added.
(WebCore::FileSystemWritableFileStream::create):
(WebCore::FileSystemWritableFileStream::FileSystemWritableFileStream):
(WebCore::FileSystemWritableFileStream::write):
(WebCore::FileSystemWritableFileStream::seek):
(WebCore::FileSystemWritableFileStream::truncate):
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.h: Copied from Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.idl.
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.idl: Copied from Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.idl.
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.cpp: Added.
(WebCore::writeParamsFromChunk):
(WebCore::fetchDataBytesForWrite):
(WebCore::FileSystemWritableFileStreamSink::create):
(WebCore::FileSystemWritableFileStreamSink::FileSystemWritableFileStreamSink):
(WebCore::FileSystemWritableFileStreamSink::~FileSystemWritableFileStreamSink):
(WebCore::FileSystemWritableFileStreamSink::write):
(WebCore::FileSystemWritableFileStreamSink::close):
(WebCore::FileSystemWritableFileStreamSink::error):
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.h: Copied from Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.idl.
* Source/WebCore/Modules/filesystemaccess/FileSystemWriteCommandType.h: Copied from Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.idl.
* Source/WebCore/Modules/filesystemaccess/WorkerFileSystemStorageConnection.cpp:
(WebCore::WorkerFileSystemStorageConnection::createWritable):
(WebCore::WorkerFileSystemStorageConnection::closeWritable):
(WebCore::WorkerFileSystemStorageConnection::executeCommandForWritable):
* Source/WebCore/Modules/filesystemaccess/WorkerFileSystemStorageConnection.h:
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
* Source/WebCore/Modules/streams/WritableStream.h:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/fileapi/Blob.cpp:
(WebCore::Blob::getArrayBuffer):
* Source/WebCore/fileapi/Blob.h:
* Source/WebCore/platform/FileHandle.cpp:
(WebCore::FileHandle::path const):
* Source/WebCore/platform/FileHandle.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::updateSharedPreferencesForWebProcess):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::createNetworkConnectionToWebProcess):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageError.h:
(WebKit::convertToException):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageError.serialization.in:
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
(WebKit::FileSystemStorageHandle::createWritable):
(WebKit::FileSystemStorageHandle::closeWritable):
(WebKit::FileSystemStorageHandle::executeCommandForWritableInternal):
(WebKit::FileSystemStorageHandle::executeCommandForWritable):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::startReceivingMessageFromConnection):
(WebKit::NetworkStorageManager::stopReceivingMessageFromConnection):
(WebKit::NetworkStorageManager::updateSharedPreferencesForConnection):
(WebKit::NetworkStorageManager::createWritable):
(WebKit::NetworkStorageManager::closeWritable):
(WebKit::NetworkStorageManager::executeCommandForWritable):
(WebKit::NetworkStorageManager::sharedPreferencesForWebProcess const):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebFileSystemStorageConnection.cpp:
(WebKit::WebFileSystemStorageConnection::createWritable):
(WebKit::WebFileSystemStorageConnection::closeWritable):
(WebKit::WebFileSystemStorageConnection::executeCommandForWritable):
* Source/WebKit/WebProcess/WebCoreSupport/WebFileSystemStorageConnection.h:

Canonical link: <a href="https://commits.webkit.org/287390@main">https://commits.webkit.org/287390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/174743ba5557c76df724d88a97dfe95769a0dbd9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79368 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58395 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32742 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83978 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30553 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81501 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67461 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6657 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62094 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19998 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82435 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52138 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72316 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42403 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49482 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26503 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28908 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72454 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70666 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26949 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85382 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78558 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6644 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4628 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70337 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6809 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68193 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69584 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17362 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13626 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12493 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100899 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6596 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24616 "Found 2 new JSC stress test failures: wasm.yaml/wasm/stress/struct-new_default-small-members.js.default-wasm, wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-no-cjit (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6510 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9985 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8308 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.cpp:
```diff
@@ -125,5 +128,58 @@ void FileSystemFileHandle::unregisterSyncAccessHandle(FileSystemSyncAccessHandle
     connection().unregisterSyncAccessHandle(identifier);
 }
 
+void FileSystemFileHandle::createWritable(const CreateWritableOptions& options, DOMPromiseDeferred<IDLInterface<FileSystemWritableFileStream>>&& promise)
```
Can we add `// https://fs.spec.whatwg.org/#api-filesystemfilehandle-createwritable`?

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.cpp:
```diff
@@ -125,5 +128,58 @@ void FileSystemFileHandle::unregisterSyncAccessHandle(FileSystemSyncAccessHandle
+void FileSystemFileHandle::createWritable(const CreateWritableOptions& options, DOMPromiseDeferred<IDLInterface<FileSystemWritableFileStream>>&& promise)
+{
+    if (isClosed())
+        return promise.reject(Exception { ExceptionCode::InvalidStateError, "Handle is closed"_s });
```
This is not as per spec, though it kind of makes sense. Worth checking what other UAs are doing.

szewai replied:
Yes, `isClosed()` is an implementation choice as we want to make the handle numb when context has stopped or connection to network process is lost -- other implementers may not have the issue. I kept the check here to match the other operations.

youennf commented on Source/WebCore/Modules/filesystemaccess/WorkerFileSystemStorageConnection.cpp:
```diff
@@ -313,6 +313,68 @@ void WorkerFileSystemStorageConnection::invalidateAccessHandle(WebCore::FileSyst
         handle->invalidate();
 }
 
+void WorkerFileSystemStorageConnection::createWritable(FileSystemHandleIdentifier identifier, bool keepExistingData, VoidCallback&& callback)
```
Another way to do this is with NativePromise, see CacheStorageConnection for instance.
One nice thing is that, if we do IPC, we could do IPC straight from the worker without going through main thread (and ditto for the answer).

szewai replied:
Yes, I thought about that, but since the patch is big, might be easier to keep the new functions consistent with existing functions, and do the refactoring for all functions later.

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.cpp:
```diff
@@ -125,5 +128,58 @@ void FileSystemFileHandle::unregisterSyncAccessHandle(FileSystemSyncAccessHandle
+
+        ExceptionOr<Ref<FileSystemWritableFileStream>> stream { Exception { ExceptionCode::UnknownError } };
+        {
+            Locker<JSC::JSLock> locker(globalObject->vm().apiLock());
```
We might want to lock inside `FileSystemWritableFileStream::create`, maybe further below in the writable stream routine.
That will also make the code more easy to read here.

szewai replied:
I thought about that, and found `WebTransport` currently acquires lock before invoking `WebTransportSendStream::create`, and I don't want to expand the change. I will leave a comment here about moving the lock inside `WritableStream` in a followup.

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.h:
```diff
@@ -0,0 +1,58 @@
+    void truncate(uint64_t size, DOMPromiseDeferred<void>&&);
+
+private:
+    FileSystemWritableFileStream(Ref<InternalWritableStream>&&);
```
explicit

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.cpp:
```diff
@@ -0,0 +1,173 @@
+void FileSystemWritableFileStreamSink::write(ScriptExecutionContext& context, JSC::JSValue value, DOMPromiseDeferred<void>&& promise)
+{
+    if (m_isClosed)
+        return promise.reject(Exception { ExceptionCode::InvalidStateError, "Sink is closed"_s });
```
Ditto here.

szewai replied:
Actually https://streams.spec.whatwg.org/#dom-underlyingsink-write specifies `write` should not be invoked after `abort()` or `close()`, so I will just change this to an `ASSERT()`.

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.cpp:
```diff
@@ -0,0 +1,173 @@
+        return promise.reject(Exception { ExceptionCode::InvalidStateError, "Sink is closed"_s });
+
+    if (value.isUndefined())
+        return promise.reject(Exception { ExceptionCode::TypeError, "Data is undefined"_s });
```
Do we have a test here? Isn't the convert routine handling the undefined case?

szewai replied:
Yes, I think `fetchDataBytesForWrite` should take care of the undefined case; will add a test.

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.cpp:
```diff
@@ -0,0 +1,173 @@
+    auto scope = DECLARE_THROW_SCOPE(context.vm());
+    auto chunkResult = convert<IDLUnion<IDLArrayBufferView, IDLArrayBuffer, IDLInterface<Blob>, IDLUSVString, IDLDictionary<FileSystemWritableFileStream::WriteParams>>>(*context.globalObject(), value);
+    if (UNLIKELY(chunkResult.hasException(scope)))
+        return promise.reject(Exception { ExceptionCode::ExistingExceptionError });
```
Do we have a test?

szewai replied:
Will add a test (wpt may have the tests but currently they all fail since this is just a partial implementation).

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.cpp:
```diff
@@ -0,0 +1,173 @@
+                return completionHandler(false, { });
+
+            Ref buffer = result.releaseReturnValue();
+            if (buffer->isDetached())
```
Can it be a case?
Maybe we can just ASSERT?

szewai replied:
I am not sure; will see if I could add a test.

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.cpp:
```diff
@@ -0,0 +1,173 @@
+    switch (writeParams.type) {
+    case FileSystemWriteCommandType::Seek: {
+        if (!writeParams.position)
+            return promise.reject(Exception { ExceptionCode::TypeError, "Position is not provided for seek"_s });
```
According to spec, we are supposed to reject asynchronously since we are going to the file system queue and then to the storage task queue.
This is observable to JS though an edge case.
I would be tempted to change the spec to check for the parameters synchronously before enqueueing tasks.

Ditto below

szewai replied:
ok I will drop the check to match spec; and file a spec issue.

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.h:
```diff
@@ -0,0 +1,51 @@
+    static ExceptionOr<Ref<FileSystemWritableFileStreamSink>> create(FileSystemFileHandle&);
+
+private:
+    FileSystemWritableFileStreamSink(FileSystemFileHandle&);
```
explicit.

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.h:
```diff
@@ -0,0 +1,51 @@
+
+private:
+    FileSystemWritableFileStreamSink(FileSystemFileHandle&);
+    ~FileSystemWritableFileStreamSink();
```
We usually make the destructors public.

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.h:
```diff
@@ -0,0 +1,51 @@
+namespace WebCore {
+class FileSystemFileHandle;
+
+class FileSystemWritableFileStreamSink : public WritableStreamSink {
```
final?

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.cpp:
```diff
@@ -0,0 +1,173 @@
+void FileSystemWritableFileStreamSink::close()
+{
+    if (m_isClosed)
+        return;
```
Do we have a test where we write then close?
It seems to me that normally, we would finish writing and then closing.
Which does not seem to be the case if we write a blob say.

szewai replied:
> Do we have a test where we write then close?

Yes, the new test does write and close.

youennf commented on Source/WebCore/Modules/filesystemaccess/WorkerFileSystemStorageConnection.cpp:
```diff
@@ -313,6 +313,68 @@ void WorkerFileSystemStorageConnection::invalidateAccessHandle(WebCore::FileSyst
+    callOnMainThread([callbackIdentifier, workerThread = Ref { m_scope->thread() }, mainThreadConnection = m_mainThreadConnection, identifier, keepExistingData]() mutable {
+        auto mainThreadCallback = [callbackIdentifier, workerThread = WTFMove(workerThread)](auto&& result) mutable {
+            workerThread->runLoop().postTaskForMode([callbackIdentifier, result = crossThreadCopy(WTFMove(result))] (auto& scope) mutable {
+                if (auto connection = downcast<WorkerGlobalScope>(scope).fileSystemStorageConnection())
```
```suggestion
                if (RefPtr connection = downcast<WorkerGlobalScope>(scope).fileSystemStorageConnection())
```
here and below

youennf commented on Source/WebCore/fileapi/Blob.cpp:
```diff
@@ -308,6 +308,13 @@ void Blob::arrayBuffer(DOMPromiseDeferred<IDLArrayBuffer>&& promise)
+void Blob::getArrayBuffer(CompletionHandler<void(ExceptionOr<Ref<JSC::ArrayBuffer>>)>&& completionHandler)
+{
+    loadBlob(FileReaderLoader::ReadAsArrayBuffer, [completionHandler = WTFMove(completionHandler)](BlobLoader& blobLoader) mutable {
+        completionHandler(arrayBufferFromBlobLoader(blobLoader));
```
This is good enough as is, since FileReaderLoader is receiving SharedBuffer, it would be nice if we could somehow write the SharedBuffer directly.
Maybe add a comment in the sink.cpp file about this future optimisation?

Or maybe, since we write in networking process, we could simply send the blob URL instead as part of the write operation.

szewai replied:
Sure, I will leave a comment in `fetchDataBytesForWrite` to say we might just send blob URL to backend.

youennf replied:
A comment sounds good, but let's not leave this for too long, given sending the blob URL allows to send synchronously, which is a nice simplification.

youennf commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -130,7 +130,7 @@ struct DataKey;
 
 class NetworkConnectionToWebProcess final
-    : public RefCounted<NetworkConnectionToWebProcess>
+    : public ThreadSafeRefCounted<NetworkConnectionToWebProcess>
```
Can we document why we need this change here? Can we do without?
Maybe we could split this patch if this is about shared preferences.

NetworkConnectionToWebProcess probably wants to be destroyed in main thread.
We could add DestructionThread::MainRunLoop to guarantee this.

szewai replied:
To do feature flag check, message handler functions need to read flag value from `SharedPreferencesForWebProcess`, which is stored in `NetworkConnectionToWebProcess`.
For `NetworkStorageManager`, the message handler functions are invoked on WorkQueue, so in this case `NetworkConnectionToWebProcess` needs to be accessed from background thread, and needs to be `ThreadSafeRefCounted`.

szewai replied:
Actually I think we could store `SharedPreferencesForWebProcess` in `NetworkStorageManager` and avoid changing `NetworkConnectionToWebProcess`. I updated.

youennf commented on Source/WebKit/WebProcess/WebCoreSupport/WebFileSystemStorageConnection.cpp:
```diff
@@ -214,6 +214,36 @@ void WebFileSystemStorageConnection::invalidateAccessHandle(WebCore::FileSystemS
     }
 }
 
+void WebFileSystemStorageConnection::createWritable(WebCore::FileSystemHandleIdentifier identifier, bool keepExistingData, WebCore::FileSystemStorageConnection::VoidCallback&& completionHandler)
```
I would be tempted to use NativePromise IPC, you can take a look at WebCacheStorageConnection.

szewai replied:
As mentioned above, we might do it as a followup for all `FileSystemStorageConnection` operations for better consistency and maintenance.

youennf commented on Source/WebKit/WebProcess/WebCoreSupport/WebFileSystemStorageConnection.cpp:
```diff
@@ -214,6 +214,36 @@ void WebFileSystemStorageConnection::invalidateAccessHandle(WebCore::FileSystemS
+    if (!m_connection)
+        return completionHandler(WebCore::Exception { WebCore::ExceptionCode::UnknownError, "Connection is lost"_s });
+
+    m_connection->sendWithAsyncReply(Messages::NetworkStorageManager::CreateWritable(identifier, keepExistingData), [completionHandler = WTFMove(completionHandler)](auto error) mutable {
```
Should we take a Ref?

szewai replied:
Take a Ref for `m_connection`?  Seem unnecessary since `sendWithAsyncReply` will not clear `m_connection`

cdumez replied:
We need to have a smart pointer to the connection on the stack, still. The safer cpp bot should complain about this unless this file is currently being ignored. Youenn is right and we should make the change no matter what.

youennf reviewed (commented):
Looks to go in the right direction. Some comments below.

youennf commented on Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.h:
```diff
@@ -0,0 +1,51 @@
+    void close() final;
+    void error(String&&) final;
+
+    Ref<FileSystemFileHandle> m_source;
```
Let's make it a `const Ref`, and then we might not need to use `protectedSource()` at all.

cdumez replied:
It doesn't hurt to do it but I doubt Ryosuke has had time to implement this in the static analyzer yet.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -287,13 +288,21 @@ void NetworkStorageManager::close(CompletionHandler<void()>&& completionHandler)
+        ASSERT(!m_preferencesForConnections.contains(connection));
+        m_preferencesForConnections.add(connection, preferences);
+
+        RunLoop::protectedMain()->dispatch([protectedThis = WTFMove(protectedThis)] { });
```
It is a bit sad we have to do this, maybe we should refactor the code a bit so that we can make sure NetworkStorageManager is DestructionThread::MainRunLoop.
We can probably do this by making WorkQueueMessageReceiver AbstractRefCounted and adding a ThreadSafe WorkQueueMessageReceiver or something like that.

szewai replied:
Good idea, I filed https://bugs.webkit.org/show_bug.cgi?id=284065.

szewai committed:
Add initial implementation of FileSystemWritableFileStream
https://bugs.webkit.org/show_bug.cgi?id=283841
rdar://problem/140709733

Reviewed by Youenn Fablet.

Add initial support for FileSystemWritableFileStream: client can now create it via createWritable() method and use it to
modify an OPFS file (via WritableStreamDefaultWriter methods) in Window context.
Spec: https://fs.spec.whatwg.org/#api-filesystemwritablefilestream.

New tests: storage/filesystemaccess/writable-file-stream-write.html
           storage/filesystemaccess/writable-file-stream-seek.html
           storage/filesystemaccess/writable-file-stream-truncate.html

Rebaseline existing tests since they can proceed futher after createWritable() is available. Some tests still fail
because the methods of FileSystemWritableFileStream are not implemented yet.

* LayoutTests/imported/w3c/web-platform-tests/file-system-access/sandboxed_FileSystemDirectoryHandle-move.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/file-system-access/sandboxed_FileSystemDirectoryHandle-move.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-IndexedDB.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-IndexedDB.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-BroadcastChannel.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-Error.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-MessagePort-frames.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-MessagePort-windows.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-MessagePort-workers.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-frames.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-windows.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-postMessage-workers.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-remove.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemBaseHandle-remove.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-getFileHandle.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-getFileHandle.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-iteration.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-iteration.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-removeEntry.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemDirectoryHandle-removeEntry.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-getFile.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-getFile.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-move.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-move.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemFileHandle-sync-access-handle-writable-lock.https.tentative.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream-piped.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream-piped.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream-write.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream-write.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/FileSystemWritableFileStream.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/idlharness.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fs/idlharness.https.any.worker-expected.txt:
* LayoutTests/storage/filesystemaccess/resources/shared.js:
(finishTest):
(async asyncReadFileAsText):
* LayoutTests/storage/filesystemaccess/resources/writable-file-stream-seek.js: Added.
(async test):
* LayoutTests/storage/filesystemaccess/resources/writable-file-stream-truncate.js: Added.
(async checkFileContent):
(async test):
* LayoutTests/storage/filesystemaccess/resources/writable-file-stream-write.js: Added.
(input.new.Number):
(async runTest):
(async test):
* LayoutTests/storage/filesystemaccess/writable-file-stream-seek-expected.txt: Added.
* LayoutTests/storage/filesystemaccess/writable-file-stream-seek.html: Added.
* LayoutTests/storage/filesystemaccess/writable-file-stream-truncate-expected.txt: Added.
* LayoutTests/storage/filesystemaccess/writable-file-stream-truncate.html: Added.
* LayoutTests/storage/filesystemaccess/writable-file-stream-write-expected.txt: Added.
* LayoutTests/storage/filesystemaccess/writable-file-stream-write.html: Added.
* Source/JavaScriptCore/runtime/GenericTypedArrayView.h:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::copyFile):
* Source/WTF/wtf/FileSystem.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.cpp:
(WebCore::FileSystemFileHandle::createWritable):
(WebCore::FileSystemFileHandle::closeWritable):
(WebCore::FileSystemFileHandle::executeCommandForWritable):
* Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.h:
* Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.idl:
* Source/WebCore/Modules/filesystemaccess/FileSystemStorageConnection.h:
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.cpp: Added.
(WebCore::FileSystemWritableFileStream::create):
(WebCore::FileSystemWritableFileStream::FileSystemWritableFileStream):
(WebCore::FileSystemWritableFileStream::write):
(WebCore::FileSystemWritableFileStream::seek):
(WebCore::FileSystemWritableFileStream::truncate):
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.h: Copied from Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.idl.
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStream.idl: Copied from Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.idl.
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.cpp: Added.
(WebCore::writeParamsFromChunk):
(WebCore::fetchDataBytesForWrite):
(WebCore::FileSystemWritableFileStreamSink::create):
(WebCore::FileSystemWritableFileStreamSink::FileSystemWritableFileStreamSink):
(WebCore::FileSystemWritableFileStreamSink::~FileSystemWritableFileStreamSink):
(WebCore::FileSystemWritableFileStreamSink::write):
(WebCore::FileSystemWritableFileStreamSink::close):
(WebCore::FileSystemWritableFileStreamSink::error):
* Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.h: Copied from Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.idl.
* Source/WebCore/Modules/filesystemaccess/FileSystemWriteCommandType.h: Copied from Source/WebCore/Modules/filesystemaccess/FileSystemFileHandle.idl.
* Source/WebCore/Modules/filesystemaccess/WorkerFileSystemStorageConnection.cpp:
(WebCore::WorkerFileSystemStorageConnection::createWritable):
(WebCore::WorkerFileSystemStorageConnection::closeWritable):
(WebCore::WorkerFileSystemStorageConnection::executeCommandForWritable):
* Source/WebCore/Modules/filesystemaccess/WorkerFileSystemStorageConnection.h:
* Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp:
* Source/WebCore/Modules/streams/WritableStream.h:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/fileapi/Blob.cpp:
(WebCore::Blob::getArrayBuffer):
* Source/WebCore/fileapi/Blob.h:
* Source/WebCore/platform/FileHandle.cpp:
(WebCore::FileHandle::path const):
* Source/WebCore/platform/FileHandle.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::updateSharedPreferencesForWebProcess):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::createNetworkConnectionToWebProcess):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageError.h:
(WebKit::convertToException):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageError.serialization.in:
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
(WebKit::FileSystemStorageHandle::createWritable):
(WebKit::FileSystemStorageHandle::closeWritable):
(WebKit::FileSystemStorageHandle::executeCommandForWritableInternal):
(WebKit::FileSystemStorageHandle::executeCommandForWritable):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::startReceivingMessageFromConnection):
(WebKit::NetworkStorageManager::stopReceivingMessageFromConnection):
(WebKit::NetworkStorageManager::updateSharedPreferencesForConnection):
(WebKit::NetworkStorageManager::createWritable):
(WebKit::NetworkStorageManager::closeWritable):
(WebKit::NetworkStorageManager::executeCommandForWritable):
(WebKit::NetworkStorageManager::sharedPreferencesForWebProcess const):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebFileSystemStorageConnection.cpp:
(WebKit::WebFileSystemStorageConnection::createWritable):
(WebKit::WebFileSystemStorageConnection::closeWritable):
(WebKit::WebFileSystemStorageConnection::executeCommandForWritable):
* Source/WebKit/WebProcess/WebCoreSupport/WebFileSystemStorageConnection.h:

Canonical link: https://commits.webkit.org/287390@main
